### ISSUE TITLE
Rewarded interstitials?

### ISSUE THREAD

**Issue by hasen6** [sbugert/react-native-admob#22]
Just wondering if the 'rewarded interstitials' work? If so how would that be implemented - the same as a normal interstitial?

**Comment by sbugert** [comment]
The API for interstitials is not quite ready yet, so there is no way to add rewarded interstitials right now.

If you have any proposals for an API don't hesistate to jump in the discussion in #6.

**Comment by hasen6** [comment]
Do you mean your plugin is not ready or the Google admob API itself is not ready? I realise its very new, it only came out about a month or so ago.

**Comment by hasen6** [comment]
But ordinary interstitials are already working aren't they? They certainly seem to be for me at least.

**Comment by sbugert** [comment]
My plugin is not ready.
Ordinary interstitials are already working on iOS and android but I want to extend the API soon.

**Comment by hasen6** [comment]
I'd love to see this implemented. Is there any way I can help? I'm not the greatest programmer but I'd be willing to do any testing or whatever you need.

**Comment by sbugert** [comment]
As far as I can tell the only difference between normal and rewared interstitials are mediation networks and the class you call the method on?

I can imagine exposing a second module in the react-native-admob package which is almost the same as the ordinary interstitials. Shouldn't be too hard! The most difficult task with this react-native module is always finding a common API for iOS and Android.

But testing would be great :) I will get in touch with you when it's ready!

**Comment by hasen6** [comment]
Ok hope you can figure it out. Really great that you provide this plugin. Certainly I can help with any testing or anything else you need me for.

**Comment by sbugert** [comment]
**Update:**
I tried setting up a rewarded ad but seems like you need a mediation network even to really test it. Do you have access to one?
With a google test id I only got a black ad with no video and whenever I add a test device id I get an error `Request Error: No ad to show`.
So I guess it would work if I had setup a network supporting rewarded ads but I'm not sure. After I publish the new interstitial API I can commit the code to another branch for you to test!

**Comment by hasen6** [comment]
You just need an account with some other Ad networks right? I haven't tried it yet but I remember I signed up to Chartboost and Flurry in the past.

**Comment by hasen6** [comment]
Btw I got everything working on Android in the end. I don't know exactly what it was but my general unfamiliarity with Android in general probably had a lot to do with it.

**Comment by hasen6** [comment]
Any progress with this...or is it a dead end?

**Comment by sbugert** [comment]
I'm quite busy these days but I think I can get something running at the end of the week or next week!

**Comment by hasen6** [comment]
I guess you must be very busy hehe.

**Comment by th0th** [comment]
Implementation of rewarded interstitials would be awesome. Any news on this?

**Comment by xing-zheng** [comment]
Same to me!

**Comment by th0th** [comment]
@sbugert I am not familiar with java, so I can't help with the coding. But I can help documenting or testing. Also, I've enabled some mediation networks on my AdMob account, I can provide any details you need.

I am so desperate that I even tried taking a look at the code but, god, java code looks complicated to me. Is there any way I can help you so that we can get rewarded videos working? :)

**Comment by hasen6** [comment]
Hey I'm the designated tester. That job is taken. :)

Btw its not Java, React is Javascript.

**Comment by th0th** [comment]
@hasen6 I just wanted to say I am ready to help when needed.

And by Java, I meant the native module for Android.

**Comment by sbugert** [comment]
I added a new branch (rewarded-video) to the repo with a first iOS version that is ready to be tested.

Of course, the more testers the better!
(Btw react is javascript, the react-native bridge is platform-specific code so either Objective-C or Java which are the two languages most of the code of this project lives)

##### Instructions:
- clone the repo
- `git checkout rewarded-video`
- `npm i` in the Example folder
- add the network adapters and [setup the mediation network](https://firebase.google.com/docs/admob/ios/mediation-networks) (they have different instructions) in the Example Xcode project
- change the ad id in **index.ios.js**

**Comment by xing-zheng** [comment]
@sbugert Great! I can't wait to test it! And I will post the feedback after test.

**Comment by xing-zheng** [comment]
Hi @sbugert , Did you forgot to check in a js file like "RNAdMobRewarded.js". I try to run the Example project follow your instructions, but It show me "Cannot read property 'setTestDeviceID' of undefined" while execute code: AdMobRewarded.setTestDeviceID('EMULATOR').

after replace 
```
import { AdMobRewarded } from 'react-native-admob';
```
with
```
const {RNAdMobRewarded:AdMobRewarded} = NativeModules;
```

It show me another error message: "AdMobRewarded.addEventListener is not a function"

**Comment by xing-zheng** [comment]
@sbugert Please review the commit #7ad8ea0

**Comment by sbugert** [comment]
You are right, I forgot the file!

Should be there now :)

**Comment by xing-zheng** [comment]
and the index.js ? :)

**Comment by sbugert** [comment]
Sorry! Should be there now :)

It's confusing because I do the development in another folder and have to copy files around.

**Comment by xing-zheng** [comment]
@sbugert Thank you for your rapid support

**Comment by th0th** [comment]
Did anybody manage to get a rewarded video displayed? I get "No ad to show" no matter I do.

**Comment by sbugert** [comment]
Have you correctly setup the mediation for the network you're using? And are rewarded ads enabled in said network?

**Comment by th0th** [comment]
@sbugert Actually, my application is for Android only. So all my accounts and such are configured that way.

I created a test iOS application and tried the account data of my android app on this new test application. I did my tests mainly on Chartboot and;

* I added Chartboost SDK to my application
* I added AdMob Chartboost mediation adapter to my application
* I configured my application's required files (as told in its docs https://answers.chartboost.com/hc/en-us/articles/201220095-Download-Integrate-the-Chartboost-SDK-for-iOS)

And when I built and run the application, I got the "no ad to show" error. At this point, I am sot sure if there is a problem with the library or Chartboost doesn't provide an ad since it expects my application to be from Android.

When I get some time, I'll try to create a new test app account for iOS on both AdMob and mediation network and try to test with the new data.

BTW, is there any progress about Android?

**Comment by th0th** [comment]
I am so excited, I got it working! I think the problem was about the iOS version I am testing. It was 7.0 (which is the XCode's default). I spent literally hours trying to integrate different SDKs from different mediation networks. A few minutes ago, a network's SDK I tried required a higher iOS version and I upgraded to 9.3 and video got loaded.

I can't wait to integrate it into my android application! @sbugert help me out :)

**Comment by hasen6** [comment]
So it only works on iOS 9?

**Comment by th0th** [comment]
I don't think it is about react-native-admob. To use rewarded videos from mediation networks, you need to integrate those networks' SDKs. I think if there is a iOS version restriction, those SDKs are the cause.

**Comment by hasen6** [comment]
So only certain networks could be incompatible.

**Comment by th0th** [comment]
I tried adding Android native module for rewarded ads trying to abide interstitial ads module. But I couldn't get it working, got this error:

> can't create handler inside thread that has not called looper.prepare()

Can someone with Java skills have a look please? Code is here:

https://github.com/th0th/react-native-admob/blob/rewarded-ads/android/src/main/java/com/sbugert/rnadmob/RNAdMobRewardedAdModule.java

**Comment by sbugert** [comment]
Great that it worked :)
I will take a look at the android version on the weekend!

**Comment by sbugert** [comment]
So, I just merged @th0th's java code into the rewarded-video branch!

Would be good if we did some more testing. @xing-zheng, did you succeed in running the iOS version yet?
Then we only need documentation and I will merge the branch and publish it to npm.

**Comment by nschurmann** [comment]
is this ready to go?? :)

**Comment by nikhilgupta90** [comment]
Is it possible to implement Rewarded Video Advertisement in React-Native for Android?

**Comment by anandvardhan1991** [comment]
Is it possible to implement Rewarded Video Ad in React-Native for Android?

If possible how to do it?

**Comment by alvaromb** [comment]
Hi! @sbugert can I help you with anything? I would love to see this merged :)

**Comment by 4nim8r** [comment]
Would love to get this working, need any help?

**Comment by alvaromb** [comment]
I'll try to release this https://github.com/sbugert/react-native-admob/pull/80 today!

**Comment by alvaromb** [comment]
Just merged #94

**Comment by alvaromb** [comment]
Published https://github.com/sbugert/react-native-admob/releases/tag/v1.2.0

Please feel free to ping me if anything is wrong.
